### PR TITLE
Add Promise/Future support

### PR DESCRIPTION
**Pull request by beckerhe**
This adds (hopefully) properly tested standard facilities to tell the caller when a background task has been finished.

ThreadPool::Schedule will return a future-type which is connected to the background job. This future can be actively
waited for by calling `MainThreadExecutor::WaitFor`. MainThreadExecutor will take care of scheduling other work on
the main thread while the wait is happening.

The future currently does not have a `Then`-method as it's typical in other programming languages. But this will come
in the future. All facilities are there to chain a main-thread-job to a background-job.

### What is the purpose of all this?

We currently have a lot of code of this structure:
```
thread_pool_->Schedule([&]() {
  // do some background work
  main_thread_executor_->Schedule([&]() {
    // implement the result of the background work in the UI.
  });
});
```

For a lot of use-cases that's fine, but it's not really extensible. In a lot of cases we want to
execute the same background job, wait for it to finish and then react to it on the main-thread.
In the previous design background-task and follow-up-task are tightly coupled. Due to the introduction
of futures, the main-thread-task can be decoupled from the background-task.

### Example

The declaration of `OrbitApp::LoadModules` currently looks like this:
```
  void LoadModules(                                                                                                     
      const std::vector<ModuleData*>& modules,                                                                          
      absl::flat_hash_map<std::string, std::vector<uint64_t>> function_hashes_to_hook_map = {},                         
      absl::flat_hash_map<std::string, std::vector<uint64_t>> frame_track_function_hashes_map = {});
```

The parameters `function_hashes_to_hook_map` and `frame_track_function_hashes_map` are basically
actions that need to happen on the main thread after asynchronously loading the modules in multiple
background threads.

This violates the separation of concern-principle and is very hard to extend. With futures I can declare
something like this:
```
orbit_base::Future<void> LoadModule(ModuleData* module);
```

This interface means the caller can decide if they want to load all modules in parallel and which task to execute
afterwards.

This particular change is not part of the PR, but will follow in another one.

### CODE REVIEW

File: src/OrbitBase/include/OrbitBase/ThreadPool.h
Comment (dimitry-) [inline src/OrbitBase/include/OrbitBase/ThreadPool.h:51]: why is it mutable?

File: src/OrbitBase/include/OrbitBase/SharedState.h
Comment (dimitry-) [inline src/OrbitBase/include/OrbitBase/SharedState.h:15]: maybe use orbit_base_internal here since these are not supposed to be used outside of Promise/Capture

File: src/OrbitBase/ThreadPoolTest.cpp
Comment (dimitry-) [inline src/OrbitBase/ThreadPoolTest.cpp:410]: nit: absl::Sleep?

File: src/OrbitBase/FutureTest.cpp
Comment (dimitry-) [inline src/OrbitBase/FutureTest.cpp:12]: In spirit of testing scenarios and not methods: rename this to Create?

File: src/OrbitBase/FutureTest.cpp
Comment (dimitry-) [inline src/OrbitBase/FutureTest.cpp:18]: and this one to MarkFinished

File: src/OrbitBase/FutureTest.cpp
Comment (dimitry-) [inline src/OrbitBase/FutureTest.cpp:27]: FinishedFutureResult?

File: src/OrbitBase/FutureTest.cpp
Comment (dimitry-) [inline src/OrbitBase/FutureTest.cpp:52]: when is Future expected to be false? maybe add a test for that too?

File: src/OrbitBase/PromiseTest.cpp
Comment (dimitry-) [inline src/OrbitBase/PromiseTest.cpp:24]: NewPromiseDoesNotHaveResult

**Review by dimitry- — COMMENTED** [review]
I am going to take another look at the CL later today, these are comments I have so far.

File: src/OrbitBase/include/OrbitBase/ThreadPool.h
Comment (beckerhe) [reply src/OrbitBase/include/OrbitBase/ThreadPool.h:51]: It mutates `promise` by calling `promise.MarkFinished()`.

File: src/OrbitBase/include/OrbitBase/SharedState.h
Comment (beckerhe) [reply src/OrbitBase/include/OrbitBase/SharedState.h:15]: Good point. I wasn't sure about our policy here. I like just adding the `_internal` suffix.

File: src/OrbitBase/include/OrbitBase/ThreadPool.h
Comment (beckerhe) [reply src/OrbitBase/include/OrbitBase/ThreadPool.h:51]: Hmm. The more I think about it. Maybe `MarkFinished` should be `const`.

Comment (beckerhe) [reply src/OrbitBase/include/OrbitBase/ThreadPool.h:51]: Circling back to this: Making `MarkFinished` const is probably also not a good idea. I guess I prefer to leave the lambda `mutable`.

File: src/OrbitBase/include/OrbitBase/SharedState.h
Comment (beckerhe) [reply src/OrbitBase/include/OrbitBase/SharedState.h]: Done.

File: src/OrbitBase/ThreadPoolTest.cpp
Comment (beckerhe) [reply src/OrbitBase/ThreadPoolTest.cpp]: Done.

File: src/OrbitBase/FutureTest.cpp
Comment (beckerhe) [reply src/OrbitBase/FutureTest.cpp]: Sure. I'm very bad at naming tests. So more feedback like that is very much appreciated!

File: src/OrbitBase/FutureTest.cpp
Comment (beckerhe) [reply src/OrbitBase/FutureTest.cpp:53]: Yes. Done.

**Comment by antonrohr** [comment]
I will look at this in detail on monday. At first glance it looks pretty awesome

File: src/OrbitBase/ThreadPoolTest.cpp
Comment (pierricgimmig) [inline src/OrbitBase/ThreadPoolTest.cpp:406]: Isn't it possible for the lambda to start executing before we release the mutex here, in which case we would lock twice? What is preventing the future from being finished by the time "EXPECT_FALSE(future.IsFinished());" is called?

Comment (pierricgimmig) [reply src/OrbitBase/ThreadPoolTest.cpp:406]: Ah nevermind, got it, the mutex inside the lambda will wait until we release it with AwaitWithTimeout

File: src/OrbitBase/ThreadPoolTest.cpp
Comment (pierricgimmig) [inline src/OrbitBase/ThreadPoolTest.cpp:410]: What is the purpose of this loop? At this point we know the lambda has been executed because it released the mutex. We are waiting for the future to be marked as finished. Why not call future.get()? Isn't there a chance that the marking as finished take more than 100ms?

File: src/OrbitBase/include/OrbitBase/Future.h
Comment (pierricgimmig) [inline src/OrbitBase/include/OrbitBase/Future.h:62]: I would have expected that calling Get() would block until the result is available like std::future<T>::get. Does this mean that the user always has to implement a loop to check IsFinished() before getting the result?

Comment (beckerhe) [reply src/OrbitBase/include/OrbitBase/Future.h:62]: Good point. I will change that to a blocking behaviour!

File: src/OrbitBase/ThreadPoolTest.cpp
Comment (beckerhe) [reply src/OrbitBase/ThreadPoolTest.cpp:410]: Yes, this part is not ideal. But this will change with your suggested behaviour change of Get(). Thanks for raising!

File: src/QtUtils/include/QtUtils/FutureWatcher.h
Comment (dimitry-) [inline src/QtUtils/include/QtUtils/FutureWatcher.h:14]: nit: maybe use // comment style everywhere? I think we use it by default

**Review by dimitry- — APPROVED** [review]
otherwise lgtm

File: src/OrbitBase/include/OrbitBase/Future.h
Comment (beckerhe) [reply src/OrbitBase/include/OrbitBase/Future.h:125]: Not yet. `Then` needed a but more thought and will follow in another PR. I hope I have the design sorted out by now. Still have to implement it though.

Comment (beckerhe) [reply src/OrbitBase/include/OrbitBase/Future.h:125]: @dimitry- Do you prefer to have it in this PR? I could also still add it.

Comment (dimitry-) [reply src/OrbitBase/include/OrbitBase/Future.h:125]: this is fine, next PR sounds good!

File: src/QtUtils/include/QtUtils/FutureWatcher.h
Comment (antonrohr) [inline src/QtUtils/include/QtUtils/FutureWatcher.h:48]: Did you consider `absl::Duration`?

File: src/OrbitBase/ThreadPoolTest.cpp
Comment (beckerhe) [reply src/OrbitBase/ThreadPoolTest.cpp:410]: Done.

File: src/OrbitBase/include/OrbitBase/Future.h
Comment (beckerhe) [reply src/OrbitBase/include/OrbitBase/Future.h:64]: Done.

File: src/OrbitGl/MainThreadExecutor.h
Comment (beckerhe) [reply src/OrbitGl/MainThreadExecutor.h]: Done.

File: src/OrbitQt/MainThreadExecutorImpl.cpp
Comment (beckerhe) [reply src/OrbitQt/MainThreadExecutorImpl.cpp]: Fixed. Thanks!

File: src/QtUtils/FutureWatcher.cpp
Comment (antonrohr) [inline src/QtUtils/FutureWatcher.cpp:50]: Above you used `timeout.has_value()`. Please be consistent

File: src/QtUtils/FutureWatcherTest.cpp
Comment (antonrohr) [inline src/QtUtils/FutureWatcherTest.cpp:132]: Maybe a test `WaitForAll({}, std::nullopt)` (empty input) is a good idea.

File: src/QtUtils/include/QtUtils/FutureWatcher.h
Comment (beckerhe) [reply src/QtUtils/include/QtUtils/FutureWatcher.h]: Good catch, I wasn't aware of that. Fixed.

**Review by antonrohr — APPROVED** [review]
Cannot wait for this to make the symbol loading code simpler! Only a few minor comments

File: src/OrbitBase/include/OrbitBase/Future.h
Comment (dimitry-) [inline src/OrbitBase/include/OrbitBase/Future.h:87]: and this.

File: src/OrbitBase/ThreadPoolTest.cpp
Comment (dimitry-) [inline src/OrbitBase/ThreadPoolTest.cpp:63]: well, that does not look good :) Maybe have Future<void> without nodiscard? I think this means we want to apply nodiscard to class and omit it for the method.

File: src/OrbitBase/FutureTest.cpp
Comment (beckerhe) [reply src/OrbitBase/FutureTest.cpp:64]: Done.

File: src/QtUtils/FutureWatcher.cpp
Comment (beckerhe) [reply src/QtUtils/FutureWatcher.cpp]: Done.

File: src/QtUtils/FutureWatcherTest.cpp
Comment (beckerhe) [reply src/QtUtils/FutureWatcherTest.cpp:132]: Good idea. Done.

File: src/QtUtils/include/QtUtils/FutureWatcher.h
Comment (beckerhe) [reply src/QtUtils/include/QtUtils/FutureWatcher.h:48]: No, I haven't. I was under the impression we prefer std-APIs over third-party-APIs.

File: src/OrbitBase/ThreadPoolTest.cpp
Comment (beckerhe) [reply src/OrbitBase/ThreadPoolTest.cpp:63]: Yes, I can do that.

Comment (beckerhe) [reply src/OrbitBase/ThreadPoolTest.cpp]: Done.

File: src/OrbitBase/include/OrbitBase/Future.h
Comment (beckerhe) [reply src/OrbitBase/include/OrbitBase/Future.h]: Ah yes. Thanks for checking.

**Comment by beckerhe** [comment]
I will wait for Pierric before rebasing and merging.

File: src/QtUtils/include/QtUtils/FutureWatcher.h
Comment (antonrohr) [reply src/QtUtils/include/QtUtils/FutureWatcher.h:48]: Okay. I don't actually know if there are benefits of absl::Duration over the std:: counterpart. I can only imagine there are some, or otherwise it would not exist. For now leave everything as is. I also dont feel strongly about it.

Comment (beckerhe) [reply src/QtUtils/include/QtUtils/FutureWatcher.h:48]: `std:chrono` has (as of C++17) no support for civil time. Just clocks, time_points and durations. `abseil` offers all of that, including time-zone support. In C++20 `std::chrono` will also get all these features and I expect abseil to slowly moving towards `std::chrono`. Don't know though if this is actually the plan.

File: src/OrbitBase/include/OrbitBase/Future.h
Comment (pierricgimmig) [reply src/OrbitBase/include/OrbitBase/Future.h]: Awesome, thanks!

**Review by pierricgimmig — APPROVED** [review]
Hi @beckerhe , thanks for the change, looks really good to me! Cheers

File: src/QtUtils/include/QtUtils/FutureWatcher.h
Comment (antonrohr) [reply src/QtUtils/include/QtUtils/FutureWatcher.h:48]: Thanks for the explanation, that makes a lot of sense.

**Pull request merged**
